### PR TITLE
multi-variable filtering

### LINKED ISSUE #9101: Multi-Variable Filtering Documentation
# Documentation: Multi-Variable Filtering for Camunda 8.10

Document the new Multi-Variable Filtering (MVF) feature shipping in Camunda 8.10 alpha3 for Operate. Phase 1 implementation is complete and merged; this issue covers creating or significantly expanding the "Filtering Process Instances" user guide section to document the new multi-variable filtering capabilities alongside existing single-variable filtering.

## Context

**Feature Epic**: [#3459 Multi-Variable Filtering](https://github.com/camunda/product-hub/issues/3459)  
**Engineering Epic**: [camunda/camunda#51790](https://github.com/camunda/camunda/issues/51790)  
**Status**: Phase 1 merged to main; feature enabled in 8.10-alpha3  
**Engineering DRI**: Eddie Tsedeke  
**Design DRI**: Zsofia Komaromi  
**Documentation Owner**: Christina Ausley

## What's Shipping (Phase 1)

- **Multi-row variable filter modal** with add/remove/edit capability
- **AND-based filtering logic** (all conditions must match simultaneously)
- **Six operators**: `equals`, `not equal`, `contains`, `is one of`, `exists`, `does not exist`
- **Nested JSON editor** with maximize-into-modal for complex values
- **Smart value transformations** (auto-quoting, comma-splitting, JSON parsing)
- **Fields ⇆ JSON switch** for advanced filtering
- **Route-based modal linkability** (URL reflects modal state)
- **Soft warning at 8+ conditions** (informational only)
- **Session persistence** of filter state via sessionStorage
- **Variable filter count indicator** in sidebar

**Not in Phase 1**: Variable name autocomplete (Phase 3 stretch goal), numeric comparison operators, nested JSON path filtering.

### Content to Create/Update

1. **Main guide page**: "Filter process instances by variables" (new or significantly expanded)
   - How to access the filter panel
   - Available filters (existing + new variable filters)
   - Single variable filters (existing capability walkthrough)
   - Multi-variable filters with AND logic (new in Phase 1)
   - Each operator explained with examples and use cases
   - Nested JSON editor walkthrough
   - Smart value transformations (documented as core feature)
   - Performance considerations (8+ conditions warning)
   - Initial interaction flow (legacy single-var inline pattern)

2. **Known limitations section**
   - Contains operator truncation (~8,000 character boundary)
   - Numeric comparison operators not supported
   - Variable name autocomplete planned as future enhancement (Phase 3)
   - Nested JSON path filtering not supported

3. **Update existing references**
   - `delete-finished-instances.md` (currently vague reference to "Filter" panel)
   - Any other pages with filter references to link to the new guide

4. **Screenshots/visual examples** (optional, based on maintainability preference)
   - Empty inline form with two stacked Name/Value inputs
   - Inline form after applying a single filter
   - Chip-list view showing 2+ conditions with "Edit conditions" button
   - Modal with multiple rows (mix of operators)
   - "contains" truncation warning in modal
   - 8-condition performance warning
   - JSON tab with record-keyed structure
   - Inline validation error example

### Key Technical Details to Document

**Smart Value Transformations** (document as core feature):

| User types | Interpreted as | API sees |
|---|---|---|
| `hello` | string | `{$eq: "\"hello\""}` |
| `42` | number | `{$eq: "42"}` |
| `true` | boolean | `{$eq: "true"}` |
| `null` | null | `{$eq: "null"}` |
| `val1, val2` (with "is one of") | array | `{$in: ["\"val1\"", "\"val2\""]}` |
| `{"x":1}` | object | `{$eq: "{\"x\":1}"}` |

**Edge cases for transformations**:
- Leading zeros (e.g., `01234`) are preserved as strings
- Comma-separated values only split with "is one of" operator; wrap in quotes to preserve literal comma
- `null` vs `"null"`: exact match is important
- `contains` operator does NOT run smart transform; searches literal substring including quotes/brackets

**Operators & Use Cases**:

| Operator | Behavior | Example |
|---|---|---|
| **equals** | Exact match on variable value | status = "pending" |
| **not equal** | Does not match the specified value | status ≠ "error" |
| **contains** | Substring match (searches ~8,000 char truncated value) | order IDs containing "2024" |
| **is one of** | Match any value in a list | priority in (high, urgent) |
| **exists** | Variable is present (any value) | specific variable is set |
| **does not exist** | Variable is not present | required variable missing |

**Contains Operator Limitation** (UI warning + doc callout):

> "contains" searches only the first ~8,000 characters of a variable value. Matches in longer values may not be returned.

**Performance Warning** (UI + doc reference):

> Filtering by many conditions can be slow. Add conditions only if you need them.

*Appears at 8+ filter conditions, informational only.*

**Initial Interaction Flow** (final design decision):
1. User selects "Variables" from filter dropdown
2. Sidebar entry "Variable Filters" appears with active filter count
3. User clicks "Add/Edit conditions" on sidebar
4. Modal opens for multi-variable filtering
5. Preserves muscle memory for existing single-variable users

## Acceptance Criteria

- [ ] Main "Filter process instances by variables" guide page created/updated and includes all sections listed under "Content to Create/Update"
- [ ] All six operators documented with clear explanations and examples
- [ ] Smart value transformations documented as core feature with table and edge cases
- [ ] Contains operator truncation limitation clearly explained with warning
- [ ] Initial interaction flow documented (legacy inline → modal escalation)
- [ ] AND logic semantics clearly explained
- [ ] Nested JSON editor workflow documented
- [ ] Existing filter references in other pages updated to link to new guide
- [ ] Known limitations section added (truncation, no numeric ops, no autocomplete)
- [ ] 8+ condition performance warning documented
- [ ] Screenshots included (if maintainability decision is favorable; optional)
- [ ] Content reviewed by engineering (Yuliia Saienko) for technical accuracy
- [ ] Content reviewed by legal/compliance team if needed
- [ ] Version targeting: 8.10+ (feature docs); 8.9 and earlier keep existing single-variable references

## Engineering Q&A Clarifications (Addressed)

✅ **Variable name autocomplete**: Document as-is; don't promise "coming soon" (blocked on unconfirmed backend API)  
✅ **Truncation warning**: Implemented in UI; exact copy provided  
✅ **Performance warning**: Implemented in UI; appears at 8+ conditions  
✅ **Smart value transformations**: Document as core feature; transformation table and edge cases provided  
✅ **Initial interaction flow**: Final design locked (legacy single-var inline pattern); document as shipping  
✅ **Cross-product consistency**: Keep docs product-specific; Test Studio alignment still in progress  
✅ **Screenshots**: Optional based on maintainability; 8 concrete UI states provided if needed  

## Related Issues & Resources

- **Product Epic**: [#3459](https://github.com/camunda/product-hub/issues/3459)
- **Engineering Epic**: [camunda/camunda#51790](https://github.com/camunda/camunda/issues/51790)
- **QA Epic**: [#3681](https://github.com/camunda/camunda-docs/issues/3681)
- **Slack Channel**: #prj-pdp-3459-multi-variable-filtering
- **Feature Flag Removal**: [camunda/camunda#55000](https://github.com/camunda/camunda/issues/55000) (Phase 1 now enabled)
- **Variable Name Discovery (Phase 3 Stretch)**: Follow-up epic to be created

## Notes for Docs Team

- **Last day of Yuliia Saienko (FE DRI)**: June 15, 2026 — she provided all engineering clarifications needed for docs (see Q&A Clarifications above)
- **Handover**: Eddie Tsedeke (Engineering DRI) will support any remaining engineering questions post-alpha3
- **Versioning**: Docs targeting 8.10+ (new feature); 8.9 and earlier remain unchanged

### PR DESCRIPTION
**Pull request by christinaausley**
## Description

Closes https://github.com/camunda/camunda-docs/issues/9101.

Comprehensive documentation for multi-variable filtering (8.10 alpha3).

**New**: `filter-process-instances.md` with operators, transformations, JSON editor, performance considerations, and best practices.

**Updated**: Sidebar, `delete-finished-instances.md`, `delete-resources.md`, `process-instance-batch-modification.md`, `process-instance-migration.md` to link to new guide.

**Pending**: Engineering review for case-sensitivity, AND indicators appearance.

## When should this change go live?

- [ ] This is a bug fix, security concern, or something that needs **urgent release support**. (add `bug` or `support` label)
- [ ] This is already available but undocumented and should be released within a week. (add `available & undocumented` label)
- [ ] This is on a **specific schedule** and the assignee will coordinate a release with the Documentation team. (create draft PR and/or add `hold` label)
- [x] This is part of a scheduled **alpha or minor**. (add alpha or minor label)
- [ ] There is **no urgency** with this change (add `low prio` label)

## PR Checklist

- [ ] The commit history of this PR is cleaned up, using [`{type}(scope): {description}` commit message(s)](https://github.com/camunda/camunda-docs/blob/main/CONTRIBUTING.MD#commit-message-header-formatting)

- [x] My changes are for **an upcoming minor release** and are in the `/docs` directory.
- [ ] My changes are for an **already released minor** and are in a `/versioned_docs` directory.

- [x] I added a DRI, team, or delegate as a reviewer for technical accuracy and grammar/style:
  - [x] [Engineering team review](https://github.com/camunda/camunda-docs/blob/main/howtos/documentation-guidelines.md#review-process)
  - [ ] [Technical writer review](https://github.com/camunda/camunda-docs/blob/main/howtos/documentation-guidelines.md#review-process) via `@camunda/tech-writers` unless working with an embedded writer.

### CODE REVIEW

File: docs/components/operate/userguide/filter-process-instances.md [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+3. Click **Add condition** to add the first variable filter.
+4. Enter the **Variable name** (free text).
+5. Select an **Operator** from the dropdown.
+6. Enter the **Value** for comparison.
+7. Click **Add condition** again to add more filters. AND indicators appear between filter rows, showing that all conditions must match.
+8. Click **Apply** to filter the instance list.
+
+The sidebar displays a count of active variable filter conditions. Click **Add/Edit conditions** again to modify existing filters.
+
+### Operators reference
```
Comment (christinaausley) [inline docs/components/operate/userguide/filter-process-instances.md:77]: Is the contains operator case-sensitive or case-insensitive?

File: docs/components/operate/userguide/filter-process-instances.md [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+### How to add multi-variable filters
+
+1. In the **Filter** panel, under **Variable Filters**, click the **Add/Edit conditions** button in the sidebar.
+2. The **Variable Filters** modal opens.
+3. Click **Add condition** to add the first variable filter.
+4. Enter the **Variable name** (free text).
+5. Select an **Operator** from the dropdown.
+6. Enter the **Value** for comparison.
+7. Click **Add condition** again to add more filters. AND indicators appear between filter rows, showing that all conditions must match.
```
Comment (christinaausley) [inline docs/components/operate/userguide/filter-process-instances.md:72]: Are AND indicators actually visible/implemented in Phase 1? If yes, what do they look like? (text "AND"? icon? line connector?)

**Review by tsedekey — CHANGES_REQUESTED** [review]
Hey @christinaausley, overall this looks great and covers all the points really well! 👍

The one thing I'd like to adjust is the wording around the ~8K character limitation. Right now we only call it out for the `contains` operator (the note in the operators table and the `:::caution` block), but the truncation actually applies to **every operator that compares against a value** — `equals`, `not equal`, `is one of`, and `contains`. Under the hood, the searchable variable value is truncated to ~8K characters in the index, and all of these operators match against that truncated value, so a match that only exists beyond the boundary won't be returned no matter which value operator is used. (`exists` / `does not exist` are unaffected, since they don't compare a value.)

Could we broaden the wording so it reads as a general limitation of value-based filtering rather than something specific to `contains`? Two concrete spots:

- **Operators reference table** — the "searches truncated varValue only" note currently sits only on the `contains` row. It'd be good to reflect that `equals`, `not equal`, and `is one of` share the same limitation.
- **`:::caution` block** — reword from "The **contains** operator searches only the first ~8,000 characters…" to something operator-agnostic, e.g.: "Variable filters search only the first ~8,000 characters of a variable value. Matches in longer values may not be returned."

One important distinction to preserve: this is about the documented behaviour. In the product UI we intentionally keep the inline warning on the `contains` operator only — people intuitively expect `contains` to scan the whole value, so flagging the truncated search there sets the right expectation, whereas for `equals` / `not equal` / `is one of` a value over ~8K characters is unusual and an explicit warning felt like noise. So please describe the limitation broadly in the docs, but let's not imply the in-product warning appears for all operators. There's a [discussion](https://camunda.slack.com/archives/C0AH075TZLN/p1782471671211719) about whether to surface the warning across all value-based operators, but the current lean is to keep the status quo.

File: docs/components/operate/userguide/filter-process-instances.md (hunk 1 above)
Comment (tsedekey) [reply docs/components/operate/userguide/filter-process-instances.md:77]: It is case-sensitive

File: docs/components/operate/userguide/filter-process-instances.md (hunk 2 above)
Comment (tsedekey) [reply docs/components/operate/userguide/filter-process-instances.md:72]: image

No, AND indicators are not visible/implemented in Phase 1

**Comment by christinaausley** [comment]
@tsedekey Thanks for the detailed review! This makes sense... I updated the docs to describe truncation as a general limitation for all value-based operators (equals, not equal, is one of, and contains). I also added that contains is case-sensitive, and I kept the product behavior distinction explicit by noting that the inline UI warning is shown on contains only.

**Review by tsedekey — APPROVED** [review]

**Review by zsofiakm — APPROVED** [review]
